cluster-autoscaler/debuggingsnapshot/debugging_snapshot.go

```diff
@@ -0,0 +1,150 @@
+	// when all the data collection is finished
+	SetEndTimestamp(time.Time)
+	// SetStartTimestamp sets the timestamp in the snapshot,
+	// when all the data collection is started
+	SetStartTimestamp(time.Time)
+	// GetOutputBytes return the output state of the Snapshot with bool to specify if
+	// the snapshot has the error message set
+	GetOutputBytes() ([]byte, bool)
+	// Cleanup clears the internal data obj of the snapshot, readying for next request
+	Cleanup()
+}
+
+// DebuggingSnapshotImpl is the struct used to collect all the data to be output.
+// Please add all new output fields in this struct. This is to make the data
+// encoding/decoding easier as the single object going into the decoder
+type DebuggingSnapshotImpl struct {
+	NodeInfo                      []*NodeInfo `json:"NodeList"`
+	UnscheduledPodsCanBeScheduled []*v1.Pod   `json:"UnscheduledPodsCanBeScheduled"`
+	Error                         string      `json:"Error,omitempty"`
+	StartTimestamp                time.Time   `json:"StartTimestamp"`
+	EndTimestamp                  time.Time   `json:"EndTimestamp"`
+}
+
+// SetUnscheduledPodsCanBeScheduled is the setter for UnscheduledPodsCanBeScheduled
+func (s *DebuggingSnapshotImpl) SetUnscheduledPodsCanBeScheduled(podList []*v1.Pod) {
+	if podList == nil {
+		return
+	}
+
+	s.UnscheduledPodsCanBeScheduled = nil
+	for _, pod := range podList {
+		s.UnscheduledPodsCanBeScheduled = append(s.UnscheduledPodsCanBeScheduled, pod)
+	}
+}
+
+// SetNodeGroupInfo is the setter for Node Group Info
+// All filtering/prettifying of data should be done here.
+func (s *DebuggingSnapshotImpl) SetNodeGroupInfo(nodeInfos []*framework.NodeInfo) {
+	if nodeInfos == nil {
+		return
+	}
+
+	var NodeInfoList []*NodeInfo
+
+	for _, n := range nodeInfos {
+		nClone := n.Clone()
+		node := nClone.Node()
+
+		nodeInfo := &NodeInfo{
+			Node: node,
+			Pods: nClone.Pods,
+		}
+
+		NodeInfoList = append(NodeInfoList, nodeInfo)
+	}
+	s.NodeInfo = NodeInfoList
+}
+
+// SetEndTimestamp is the setter for end timestamp
+func (s *DebuggingSnapshotImpl) SetEndTimestamp(t time.Time) {
+	s.EndTimestamp = t
+}
+
+// SetStartTimestamp is the setter for end timestamp
+func (s *DebuggingSnapshotImpl) SetStartTimestamp(t time.Time) {
+	s.StartTimestamp = t
+}
+
+// GetOutputBytes return the output state of the Snapshot with bool to specify if
+// the snapshot has the error message set
+func (s *DebuggingSnapshotImpl) GetOutputBytes() ([]byte, bool) {
+	defer s.Cleanup()
+	errMsgSet := false
+	if s.Error != "" {
+		klog.Errorf("Debugger Found with error message set when GetOutputBytes() is called. - ", s.Error)
+		errMsgSet = true
+	}
+
+	klog.Infof("debug snapshot flush ready")
+	marshalOutput, err := json.Marshal(s)
+
+	// this error captures if the snapshot couldn't be marshalled, hence we create a new object
+	// and return the error message
+	if err != nil {
+		klog.Error("Unable to json marshal the debugging snapshot,", err)
```

Comment:
nit: missing space at the end of string. Also, the usual convention is to use a colon instead of a comma before embedding underlying error.